MSON Namespace origin

zdne opened the pull request:
This PR introduces MSON namespace – a minimal set of elements needed to replace existing [MSON AST](https://github.com/apiaryio/mson-ast) with the element-based MSON DOM. 

The Namespace is split into two parts: 
1. General used elements – elements that, in my opinion, may be used outside of the MSON domain
2. MSON DOM specific elements.

Please pay close attention to the `Distinct Element` as I believe this is the "future proof" element type concept that enables us to do referencing including namespaces. Note it is out of the scope of this specification to solve the namespaces of element identifiers. The `Distinct Element` element, however, is designed with future element identifier namespaces in mind. It should be possible to solve it by adding `namespace`, `prefix` and possibly `class` attributes to the `Distinct Element` attributes. I am open to discuss this as part of this PR, but it is not as important as providing SDKs based on this new MSON DOM structure.

Commit:
MSON namespace WIP

Commit:
Man, this really clicks nicely

Commit:
Add enum example

Commit:
Finalize MSON namespace

Commit:
Fix one of example

Commit:
Add content

Commit:
Add MSON namespace

smizell commented:
I'm really liking this stuff within the MSON namespace. Really cool, and will be diving in more, but makes sense to me.

Related to this how would one convert this into idiomatic JSON? Reason being, there is fundamentally very little difference between the representation of an array and an object. Will an element always reference back to an MSON primitive? If so, it seems like a library will have to tell elements how they should represent themselves if they are a property or an entire object themselves. 

Not sure of the right way to think about that, which is why I'm mentioning it. Thoughts?

smizell commented:
cc: @fosrias

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+# MSON Namespace
+
+This document extends [Refract][]'s [JSON Namespace][] with elements necessary to
+build [MSON][] DOM.
```
build an [MSON] ...

fosrias replied:
Do you really want to call this a DOM?

zdne replied:
Well I quite like it. Since it actually is DOM. Isn't ? Frankly over the years what I have learned is that AST is sort of alien to many, but everybody has a notion of what a DOM is. 

Is there any bad connotation to it? I don't know. 

I also like the "MSON refract" term you suggest at the bottom of this review, however, I would say I would much rather call these "namespaces" refracts (instead of namespaces)...

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+# Terminology
+
+This specification uses terminology from [MSON Specification][], in addition, bellow you can find the explanation of terms defined in this namespace.
```
This specification uses terminology from [MSON Specification][]. In addition, the terms defined in this namespace are described below.

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+## Expanded Element
+
+MSON is built around the idea of defining recursive data structures. To provide abstraction, for convenience reasons and to not repeat ourselves these structures can be named (using an _identifier_) and reused. In MSON, the reusable data structures are called _Named Types_.
```
repeat ourselves,

In [MSON][], reusable ...

or

In [MSON][], these reusable ...

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+MSON is built around the idea of defining recursive data structures. To provide abstraction, for convenience reasons and to not repeat ourselves these structures can be named (using an _identifier_) and reused. In MSON, the reusable data structures are called _Named Types_.
+
+Often, before an MSON DOM can be processed referenced _Named Types_ have to be resolved. Resolving references to _Named Types_ is tedious and error prone. As such parser can resolve references and produce a complete MSON DOM, that is a DOM that does not include unresolved references to other data structures. This is referred to as _reference expansion_ or simply _expansion_.
```
referenced, _Named Types_ ...

Do you want to add a link to MSON Named Types spec section? Here and elsewhere?

zdne replied:
Yeah if this goes through the "first reading" I think it can be augmented with additional links to MSON spec terms (and intra this document too).

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+Often, before an MSON DOM can be processed referenced _Named Types_ have to be resolved. Resolving references to _Named Types_ is tedious and error prone. As such parser can resolve references and produce a complete MSON DOM, that is a DOM that does not include unresolved references to other data structures. This is referred to as _reference expansion_ or simply _expansion_.
+
+In other words, an expanded element is such an element that does not contain _Identifier_ (defined bellow) to any other elements but those defined in JSON or MSON namespaces.
```
In other words, an expanded element is one that does not contain any _Identifier_ (defined below) referencing  any other elements than those defined in JSON or MSON namespaces.

Do we want to keep in this 'expanded' object the original reference metadata in some src or whatever attribute? You advocated having the two, but anything that may need to correlate the original and expanded versions would really be a pain to do, I suspect. No use cases in mind, but correlating will rear its ugly head somewhere.

zdne replied:
I do propose the "ref" attribute (element) to take care of this. I actually think it might work. For example when you are rendering a documentation from such a DOM you may stop at first element with "ref" and display name (id) from that ref, knowing that it will be defined somewhere else...

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+In other words, an expanded element is such an element that does not contain _Identifier_ (defined bellow) to any other elements but those defined in JSON or MSON namespaces.
+
+The expanded DOM MUST, however, keeps the track of what data structure was expanded and what from where.
```
keep

Derp. So, you say it must keep track per my prior comment, but you don't say how. I will read on, but this should be explicitly spelled out. Vague here. Put reference to other section if described there.

zdne replied:
Yeah, you are right the revelation is in the `Distinct Element` definition. This was done ex-post to sort of elaborate on terms used in the `Distinct Element` definition...

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+## Base Element
+
+In MSON, every data structure is a sub-type of another data structure, and, in transition, it is directly or indirectly derived from one of the MSON _Base Types_. This is expressed as an inheritance of elements in MSON DOM. Where the predecessor of an element is referred to as element's _Base Element_.
```
> in transition, it is directly or indirectly derived 

Don't know what you mean by this

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+In MSON, every data structure is a sub-type of another data structure, and, in transition, it is directly or indirectly derived from one of the MSON _Base Types_. This is expressed as an inheritance of elements in MSON DOM. Where the predecessor of an element is referred to as element's _Base Element_.
+
+Note not every MSON _Base Type_ is presented in JSON namespace primitive types and vice versa, see the table bellow:
```
Note: Not every ...

fosrias commented on namespaces/mson-namespace.md:
````diff
@@ -0,0 +1,503 @@
+which is identical to:
+
+```json
+{
````
First WTF here. Not sure I am groking this.

fosrias replied:
Feels like a dissonant overloaded use of id here. This is really confusing. I think the thing that creates the dissonance is the ID Element.

zdne replied:
So you are going to `WTF`-me ? :) 

Well, if you would read the MSON you would see that identifier value either `(string)` or `(id element)`

The former is handy when there is not ambiguity, once you need to dis-ambiguate the string value you need to use full-blown element (that may eventually have an namespace or prefix in it).

``` json
{
  "a-property-with-identifier-value": "identifier"
}
```

if "identifier" is ambiguous you need to elaborate, for example (out of scope of this spec): 

``` json
{
  "a-property-with-identifier-value": {
    "element": "id",
    "attributes": {
         "namespace": "lorem ipsum"
    }
    "content": "identifier"
  }
}
```

smizell replied:
Just a note, @zdne, that this is the place too that I don't really get :) I'm still trying though.

fosrias replied:
@zdne Actually, I think I see what you are doing here and this is something that @smizell and I are going to look into w/r to a `type` base element in the base spec that simplifies and fixes a number of things. Will keep you posted.

fosrias commented on namespaces/mson-namespace.md:
````diff
@@ -0,0 +1,503 @@
+
+```json
+{
+    "element": "my-number"
````
Next WTF here. Seems to me there is a fundamental spec (around the namespacing one probably) that dictates referencing and re-use that _Named Types_ should then leverage. That is just my gut feeling so far looking at this.

fosrias replied:
That being said, I think we could at least have the rule that if something like `my-number` dictates looking in current doc and then up through namespaces. But the `expansion` here is messing with me a bit.

zdne replied:
> . Seems to me there is a fundamental spec (around the namespacing one probably) that dictates referencing and re-use that Named Types should then leverage. 

There is none :( This acts instead of it.

zdne replied:
> That being said, I think we could at least have the rule that if something like my-number dictates looking in current

This is correct

> and then up through namespaces

there are no namespaces in consideration in this spec

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+{
+    "element": "number",
+    "attributes": {
+        "ref": "my-number"
```
I think this definition belongs in the namespace/referencing spec personally.

zdne replied:
OK I see the confusion. This space took care both of referencing AND MSON since there was no referencing spec (and that @smizell originally PR-ed didn't work for me) so I have suggested I will propose a spec that will combine BOTH referencing AND MSON because it might be easier to demonstrate the referencing need on actual spec that needs it.

Eventually they should be separate, yes.

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+- `element`: id (string, required, fixed)
+- `content` (string, required)
+
+## Distinct Element (Element)
```
Though I think this should make over into the namespace spec, I wonder if defining this as a `Reference-able Element` might be more ubiquitous.

zdne replied:
see my comment above

pksunkara replied:
What is the `Element` here? Do you mean `Refract Element`? If not, where is the definition.

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+- `element`: select (string, required, fixed)
+- `content` (array[Option])
+
+## Option (Distinct Element)
```
Couldn't this just be implicit that the elements of the contents of a select element are the options?

``` json
{
    "element": "select",
    "content": [
         {
                  "element": "string",
                  "content": "volvo"
              },
                {
                    "element": "string",
                    "content": "Saab"
                }
            ]
}
```

fosrias replied:
Lazy on formatting :p

zdne replied:
> Couldn't this just be implicit that the elements of the contents of a select element are the options?

No because there may be nested or grouped element, without `option` you couldn't tell whether its another "option" or a "group"

``` html
<select>
  <option>
    <a></a>
    <p>A</p>
  </option>
  <option>
    <p>B</p>
  </option>
</select>
```

vs. 

``` html
<select>
  <a></a>
  <p>A</p>
  <p>B</p>
</select>
```

pksunkara replied:
Can you give me an example on how the second snippet in the above example be represented?

zdne replied:
@pksunkara that was just an example to demonstrate the need for the `option` grouping element, the second example isn't meant to be a valid one.

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+    - `validation` - Not used, reserved for a future use
+    - `description` - Combined description of an MSON Element
+
+## Boolean Type (JSON:Boolean Type)
```
If these just extend another simple primitive types (even in the base spec), what is the value of namespacing these here? @smizell ditto for bringing definitions into the JSON namespace vs. just using the refract primitives as definitions.

zdne replied:
Thats just a technical think I have made up to say that I am extending JSON types, yes.

smizell replied:
Just a note, there are no primitives in the base spec anymore, as they were pulled out to the JSON namespace.

fosrias replied:
@smizell Yeah, well I was sort of questioning that move too here.

pksunkara replied:
@zdne Maybe I missed something but I didn't get the `JSON:Boolean Type` until I realized that it is a namespace.

zdne replied:
> Maybe I missed something but I didn't get the JSON:Boolean Type until I realized that it is a namespace.

Yup this spec should / will be decorated with link to respective namespaces / MSON spec

fosrias commented on namespaces/mson-namespace.md:
````diff
@@ -0,0 +1,503 @@
+    - green
+```
+
+#### MSON DOM
````
What if we start using the term "refracted". That is, instead of MSON DOM, say something like _Refracted MSON_ or _Refracted JSON_ or _Refracted XML_. Just a crazy thought @zdne @smizell.

smizell replied:
Ha, that's actually pretty cool.

On Thu, Feb 12, 2015 at 4:26 PM, Mark W. Foster notifications@github.com
wrote:

> In namespaces/mson-namespace.md
> https://github.com/refractproject/refract-spec/pull/4#discussion_r24626809
> :
> 
> > +### Properties
> > +
> > +- `element`: enum (string, required, fixed)
> > +- `content` (array[MSON Element])
> > +
> > +### Example
> > +
> > +#### MSON
> > +
> > +```
> > +- tag (enum[string])
> > -    - red
> > -    - green
> >   +```
> >   +
> >   +#### MSON DOM
> 
> What if we start using the term "refracted". That is, instead of MSON DOM,
> say something like _Refracted MSON_ or _Refracted JSON_ or _Refracted XML_.
> Just a crazy thought @zdne https://github.com/zdne @smizell
> https://github.com/smizell.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/refractproject/refract-spec/pull/4/files#r24626809.

zdne replied:
> Refracted MSON or Refracted JSON or Refracted XML.

Sounds cool. However I am certain nobody will understand it ... (see my comment above)

fosrias replied:
@zdne Of course you are right on the understanding part. Is this always tied to the concept of a 'document' so that there is a document object model around? If you want to go the DOM route, I think you should define it explicitly vs. assume and understanding since it's formal definition is tied to html and xml as an API: http://www.w3.org/TR/WD-DOM/introduction.html

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+- `attributes`
+    - `typeAttributes` (array) - _Type Definition_ attributes list, see _Type Attribute_  
+        - (enum[string])
```
Shouldn't this be (enum[string], fixed)

zdne replied:
Since you weren't there when we made MSON spec: No, enums are treated as fixed by default.

fosrias replied:
@zdne Hey, be nice.

fosrias commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+            - fixed
+            - sample
+            - default
+    - `variable` (boolean) - Element content is _Variable Value_
```
fixed here and in following too, no?

zdne replied:
What you think fixed does? :D

fosrias replied:
It makes the sample value fixed. That is that these are the actual attribute names instead of just samples.

fosrias commented on namespaces/mson-namespace.md:
````diff
@@ -0,0 +1,503 @@
+
+```json
+{
+    "element": {
````
I think expanding this here really confuses things and I am trying to see if this is technically correct. But I think whatever going on right here, and getting it right is the secret sauce. Hasn't completely converged for me. We you were here and we all were at the whiteboard.

fosrias commented:
Ok. So this is a great start. I think there are some ideas here that don't belong here. I think I get what you are grappling with. Going to let this percolate a bit.

Commit:
Update per review comments

zdne commented:
@fosrias typos / clarification updated & pushed. Added comments, hopefully explaining some stuff.

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+This namespace defines following elements:
+
+1. General-use elements
+    1. [Id Element](#id-element-distinct-element)
```
Wouldn't it better if this is properly named as `Identifier Element`?

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+    1. [Option](#option-distinct-element)
+
+1. MSON DOM-specific elements
+    1. [MSON Element](#mson-element-distinct-element)
```
Please convert all these links to `[Text][]` format.

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+General elements defined inside MSON namespaces but possibly reusable in another domain.
+
+## Identifier (Enum)
```
The super-types are not linked. Similarly below.

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+- `element` (Identifier, required) - identifier of this element's base element
+- `attributes`
+    - `id` (Identifier) - identifier of this element
+    - `ref` (Identifier) - reference to expanded element
```
Add me to the list of people who don't get this. What do you mean by `expanded element`? Going by the below examples, it looks like we can use `element` to denote a base element. Then, why do we need `ref`?

zdne replied:
`ref` is needed to keep the reference to the original element from whom the content of this element has been expanded. See expanded element definition on top of this document.

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+## Boolean Type (JSON:Boolean Type)
+
+- Include MSON Element
```
You might want give examples here like how you did with the `Option Element`

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+
+### Properties
+
+- `element`: enum (string, required, fixed)
```
We don't need `required` here.

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+        {
+            "element": "property",
+            "attributes": {
+                "name": "id",
```
I just realised there is no mention of this `attributes.name` in the spec except for the examples.

pksunkara replied:
Did you mean to replace `name` by this? https://github.com/refractproject/refract-spec/pull/4/files#diff-925248c117ee875155bdb1606d4fd897R90

zdne replied:
The `name` This comes from the JSON namespace – https://github.com/refractproject/refract-spec/blob/master/namespaces/json-namespace.md#property-type-refract-element

pksunkara replied:
Then, why do we need `id` key? Sorry, this is a bit confusing for me until I can see API Blueprint AST.

zdne replied:
- id key defines and identifier for an element
- ref key specifies from what identifier was the element expanded

pksunkara commented on namespaces/mson-namespace.md:
```diff
@@ -0,0 +1,503 @@
+            }
+        },
+        {
+            "element": "select",
```
I understand this, but in the spec, there is no where any mention of `content` containing `Select` named type.

zdne replied:
Good point I will update this accordingly.

Commit:
Object type allows Select element

The content of the object element should allow select element in
addition to property element

Commit:
WIP: update to latest Refract spec

Commit:
WIP: Improved support of new Refract spec

zdne commented:
Superseded by #7
